docs/source/usage.md

```diff
@@ -67,23 +67,57 @@ You can then use, e.g.,
 
 ### IPython magic
 
-In IPython (and therefore in Jupyter), you can directly execute Julia
-code using `%%julia` magic:
+In IPython (and therefore in Jupyter), you can directly execute Julia code using `%julia` magic:
 
-```
+```python
 In [1]: %load_ext julia.magic
 Initializing Julia interpreter. This may take some time...
 
-In [2]: %%julia
-   ...: Base.banner(IOContext(stdout, :color=>true))
-               _
-   _       _ _(_)_     |  Documentation: https://docs.julialang.org
-  (_)     | (_) (_)    |
-   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
-  | | | | | | |/ _` |  |
-  | | |_| | | | (_| |  |  Version 1.0.1 (2018-09-29)
- _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
-|__/                   |
+In [2]: %julia [1 2; 3 4] .+ 1 
+Out[2]: 
+array([[2, 3],
+       [4, 5]], dtype=int64)
+```
+
+You can "interpolate" Python objects into Julia code via `$var` for the value of single variables or `py"expression"` for the result of any arbitrary Python code:
+```julia
+In [3]: arr = [1,2,3]
+
+In [4]: %julia $arr .+ 1
+Out[4]: 
+array([2, 3, 4], dtype=int64)
+
+In [5]: %julia sum(py"[x**2 for x in arr]")
+Out[5]: 14
+```
+
+Python interpolation is not performed inside of strings (instead this is treated as regular Julia string interpolation), and can also be overridden elsewhere by using `$$...` to insert a literal `$...`:
+
+```julia
+In [6]: %julia foo=3; "$foo"
+Out[6]: '3'
+
+In [7]: %julia bar=3; :($$bar)
+Out[7]: 3
```

Comment:
I think the behavior comparable to PyCall is

```
In [6]: bar = "3"
   ...: %julia $$bar ^ 4
Out[6]: "3333"

In [7]: %julia $$foo ^ 4
Out[7]: 81
```

See:

```julia
julia> bar = "3"
       py"$bar * 4"
"3333"

julia> bar = "3"
       py"$$bar * 4"
12
```

Reply:
Did you have an extra `$` in your In[6]? I think that first case is totally consistent between Python/Julia currently: 

```
julia> bar = "3"
       py"$bar * 4"
"3333"

In [15]: bar = "3" 
    ...: %julia $bar ^ 4                                                                                                            
Out[15]: '3333'
```

The other case is indeed different because PyCall chooses to have `$$foo` mean interpolate foo as string into the Python expression, whereas I chose to have it mean a literal `$foo`. I think it makes some sense these can be different because `$foo` is valid Julia but not Python syntax. In any case, I guess the usecase of `py"$$foo"` which is to use Julia to build Python expressions seems really unlikely to be used in the other direction, whereas wanting a literal `$foo` seems much more common, so I'd be apt to keep it as is. Thoughts?

Reply:
> Did you have an extra `$` in your In[6]?

Oops. Good catch. I meant to write `%julia $bar ^ 4`. Fixed now.

Reply:
> wanting a literal `$foo`

it can be done with

```
foo = "$foo"
%julia $$foo
```

though I guess this is a bit strange.

The literal `$foo` in Julia is valid only in macro and quote contexts. I think we need to stop replacing `$` and `$$` inside those contexts.  We can still use `py"..."` everywhere so I think it's OK.

Reply:
I think I'll have to think about this one over the weekend. My issue with stopping replacing `$` and `$$` in macros and quotes is two fold:

1) not all macros expand `$foo`, e.g. `@eval $foo` does but `@time $foo` does not, so I don't think it makes sense to stop expanding in all macros, and it may be hard/impossible to detect which ones do expand
2) in quotes, what about `:($($foo))` where logically `$foo` should be interpolated? you could do this, but then you'd have to track the "quote depth" which would be more complex.

For these reasons I thought the current rule was both simpler to remember and to code up. Then again, I guess the depth thing *is* currently handled in string interpolation (because Julia did for me for free). I.e. `%julia "$($foo)"` interpolates `foo` from Python. So maybe it would make most sense to work in expressions too so that `%julia :($($foo))` worked and interpolated `$foo`...

Reply:
> For these reasons I thought the current rule was both simpler to remember and to code up.

I argue that a _valid_ Julia code snippet has to be evaluated correctly in `%%julia` magic [1].  Assigning `$$` a different meaning in `%%julia` violates this.  I think it is the best to make `%%julia` a _superset_ of the Julia language; i.e., additionally allow `$foo` and `$$bar` in the "toplevel" contexts which is not the case in standard Julia.  Simple to use does not mean simple to implement.  I'm all for having a few AST-manipulating functions to help users.

> 1. not all macros expand `$foo`, e.g. `@eval $foo` does but `@time $foo` does not

I'd say we should play on the safe side.  It's impossible to list all cases where `@somemacro $foo` is valid or not.  So it's safer to stop touching the AST as soon as we see a macro call.

> 2\. in quotes, what about `:($($foo))` where logically `$foo` should be interpolated?

Yes, this is a good point.  We better do this.  But I think leaving it to throw an error at this point is OK, too.  Users' code will be forward compatible to new PyJulia supporting it.

Maybe the easiest thing here is to drop `$foo` from _this_ PR and do it afterwards?  We can still use `py"..."` for all contexts, right?  It is always much easier to change the syntax that was an error to do something useful than change the behavior that breaks users' code.

[1] Side note: Local `py"..."` does not violate this because it would not be valid in vanilla Julia + PyCall.  But maybe this point is a bit subtle.